block-layer/codecs/dag-cbor.md

```diff
@@ -4,38 +4,45 @@
 
 DagCBOR supports the full [IPLD Data Model].
 
-CBOR already natively supports all [IPLD Data Model Kinds].
+DagCBOR uses the [Concise Binary Object Representation (CBOR)] data format, which natively supports all [IPLD Data Model Kinds].
 
 ## Format
 
-The CBOR IPLD format is called DagCBOR to disambiguate it from regular CBOR.
-Most CBOR objects are valid DagCBOR. The only hard restriction is that any field
-with the tag 42 must be a valid CID.
+The CBOR IPLD format is called DagCBOR to disambiguate it from regular CBOR. Most CBOR objects are valid DagCBOR. The primary differences are:
+  * tag `42` interpreted as CIDs
+  * maps may only be keyed by strings
+  * additional strictness requirements about valid data encoding forms
 
-## Link Format
+## Links
 
-As with all IPLD formats, DagCBOR must be able to encode [Links].
-In DagCBOR, links are [CIDs] encoded using the raw-binary identity [Multibase]. That Multibase prefix (`0x00`) *must not* be omitted. They are stored as byte-string type (major type 2), with the tag 42.
+As with all IPLD formats, DagCBOR must be able to encode [Links]. In DagCBOR, links are the binary form of a [CID] encoded using the raw-binary identity [Multibase]. That is, the Multibase identity prefix (`0x00`) is prepended to the binary form of a CID and this new byte array is encoded into CBOR as a byte-string (major type 2), with the tag `42`.
 
-(the inclusion of the Multibase exists for historical reasons)
+The inclusion of the Multibase prefix exists for historical reasons and the identity prefix *must not* be omitted.
 
-## Map Key Restriction
+## Map Keys
 
-In DagCBOR, map keys must be strings.
+In DagCBOR, map keys must be strings, as defined by the [IPLD Data Model].
 
-## Canonical DagCBOR
+## Strictness
 
-Canonical DagCBOR must:
+DagCBOR requires that there exist a single way of encoding any given object, and that encoded forms contain no superfluous data that may be ignored or lost in a round-trip decode/encode.
 
-1. Use no tags other than the CID tag (42). Other tags may be lost in
-   conversion.
-2. Use the [canonical CBOR](https://tools.ietf.org/html/rfc7049#section-3.9)
-   encoding.
+Therefore the DagCBOR codec must:
+
+1. Use no tags other than the CID tag (`42`). A valid DagCBOR encoder must not encode using any additional tags and a valid DagCBOR decoder must reject objects containing additional tags as invalid.
+2. Use the canonical CBOR encoding defined by the the suggestions in [section 3.9 of the CBOR specification]. A valid DagCBOR decoder should reject objects not following these restrictions as invalid. Specifically:
+   * Integers must be as small as possible.
```

Comment:
@rvagg this is somewhat ambiguous. Do you mean `Integer encoding must be as short as possible`, just like in the next item ?

Reply:
tag 98 for COSE signing may be helpful.  before merging can we discuss?  https://github.com/ipld/specs/issues/227#issuecomment-577432065

Reply:
also, see the updated draft of irc 7049: https://tools.ietf.org/html/draft-ietf-cbor-7049bis-12#section-4.2.2 regarding "Additional Deterministic Encoding Considerations" for CBOR

Reply:
@jonnycrunch re COSE seems like a larger concern and perhaps even touches on the CID spec, would you mind deferring that to a separate issue where we can all wrap our heads around it - either open a fresh issue just for it, making the case and giving us the resources we need to grok it fully, or even open a PR against this doc if you have a firm proposal for how it would work and how encoders & decoders would need to interact with it.

Reply:
re "Additional Deterministic Encoding Considerations", I did kind of ignore that but probably shouldn't have, the main area it touches on is floats, I don't know what our current CBOR encoders are doing but we should probably opt for "Encode all values as the smallest of 16-, 32-, or 64-bit floating point that accurately represents the value, even for integral values", that's consistent with the int rule. I just don't know whether that's going to work well across languages, we might end up with inconsistencies here and need additional floating point rules. The alternative is to put all floats into 64-bit.

Another thing to consider is how to handle very large integers, or whether they can be handled. We can probably hand-wave that for now like we do in the data model but it'll likely come up again in the future.

Reply:
IIRC the issue with "smallest encoding that accurately represents the value" is actually kind of an ambiguous -- or at least highly detailed, shall we say -- minefield with floats.  Ints?  Clear.  Floats?  Ngh.

The go implementations [currently make all floats 64bit](https://github.com/polydawn/refmt/blob/3d65705ee9f12dc0dfcc0dc6cf9666e97b93f339/cbor/cborEncoderTerminals.go#L69-L77), fwiw.  I wouldn't be opposed to trying to improve this, but I *would* be terrified, personally, if that makes sense.

Reply:
Ah, I just saw the [link to borc](https://github.com/dignifiedquire/borc/blob/b6bae8b0bcde7c3976b0f0f0957208095c392a36/src/encoder.js#L133-L149) in your https://github.com/ipld/specs/pull/236#issuecomment-578027088 .

I'd be fine taking patches to change the go libraries to do the same, if we're sure that logic is correct and we want to stand with it.  (I just don't feel qualified to actually review that bitshuffling logic, myself.)

Reply:
perhaps it could just be worded: `the default encoding for MIME type application/ipld+dagcbor ignores all other tag other than tag 42`.  Unfortunately, this get into the MIME type discussion. But, I could work with that if I can generate an application/ipld+cbor which is the full cbor spec with all tags.  to be clear application/ipld+dagcbor only support tag 42 whereas application/ipld+cbor is all tags.

Reply:
see: https://github.com/ipld/specs/issues/102

Reply:
a major part of the exercise here is, as stated in the changes:

>  DagCBOR requires that there exist a single way of encoding any given object, and that encoded forms contain no superfluous data that may be ignored or lost in a round-trip decode/encode.

If we _ignore_ rather than _disallow_ then we make significant space for arbitrary data to be included in DagCBOR blocks, there's no "just one way to encode" guarantee.

This is only for IPLD DagCBOR (multicodec `0x71`), what an IPLD CBOR (multicodec `0x51`) does is another matter, that can be the flexible space while DagCBOR is what you use if you want these strictness guarantees. Maybe that's something we should hasten the development, it seems like low-hanging fruit to make space for much more interesting data importing and linking.